Persistence/Repos/IBank.cs

```diff
@@ -0,0 +1,177 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Persistence.Models;
+
+namespace Persistence.Repos
+{
+    /// <summary>
+    /// Exception thrown when a user related operation failed because the user did not exist.
+    /// </summary>
+    public class UserNotFoundException<T> : Exception
+    {
+        public T User { get; }
+
+        public UserNotFoundException(T user)
+        {
+            User = user;
+        }
+    }
+
+    /// <summary>
+    /// A single atomic monetary transaction that can be performed.
+    /// </summary>
+    /// <typeparam name="T">User object type the transaction is performed on, typically <see cref="User"/></typeparam>
+    public readonly struct Transaction<T>
```

Comment:
might it be an idea to have some inheritance for the types of users, rather than just passing in a T here

Reply:
why? what's bad about being agnostic about what and what isn't a "user" as long as it meets the requirements?

Reply:
well people can pass in objects that don't necessarily meet the requirements, even if it's just a constraint `where T : IUser` or something

Reply:
I tried replacing the idFieldAccessor and currencyFieldAccessor parameters in the bank class with an interface:
```c#
public interface IBankUser
{
    public string Id { get; }
    public int Currency { get; protected internal set; }
}
```
that works, and makes the Bank implementation a lot easier, but then you cannot have banks on the same class but different currency fields anymore. That would require some sort of proxy classes like:
```c#
class PokeyenProxy : IBankUser
{
    private readonly User User;
    public PokeyenProxy(User user) => User = user;

    public string Id => User.Id;
    int IBankUser.Currency
    {
        get => User.Pokeyen;
        set => User.Pokeyen = value; // doesn't even compile because the setter is private
    }
}
```

Reply:
imo not worth it